File: js/test/prefetch.test.js
// Prefetch unit tests — run: node test/prefetch.test.js

import {
  coalescePageIndices,
  PageCache,
  InFlightMap,
  DEFAULT_PAGE_SIZE,
  HINT_FULL,
} from "../prefetch.js";
import { AccessPatternDetector } from "../pattern.js";
import { NxsReader } from "../nxs.js";
import { NxsSchema, NxsWriter } from "../nxs_writer.js";

let passed = 0;
let failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`  ✓ ${name}`);
    passed++;
  } catch (e) {
    console.log(`  ✗ ${name}\n      ${e.message}`);
    failed++;
  }
}

async function testAsync(name, fn) {
  try {
    await fn();
    console.log(`  ✓ ${name}`);
    passed++;
  } catch (e) {
    console.log(`  ✗ ${name}\n      ${e.message}`);
    failed++;
  }
}

function buildRecords(n) {
  const schema = new NxsSchema(["id", "username", "score", "active"]);
  const w = new NxsWriter(schema);
  for (let i = 0; i < n; i++) {
    w.beginObject();
    w.writeI64(0, i);
    w.writeStr(1, `user_${i}`);
    w.writeF64(2, i * 0.25);
    w.writeBool(3, i % 2 === 0);
    w.endObject();
  }
  return w.finish();
}

console.log("\nNXS Prefetch — Tests\n");

test("pattern unknown until min observations", () => {
  const d = new AccessPatternDetector();
  for (let i = 0; i < 8; i++) d.observe(i);
  if (d.pattern() !== "unknown") throw new Error(`expected unknown, got ${d.pattern()}`);
  d.observe(8);
  if (d.pattern() === "unknown") throw new Error("expected classified pattern after 9th access");
});

test("pattern sequential small deltas", () => {
  const d = new AccessPatternDetector();
  for (let i = 0; i < 20; i++) d.observe(i);
  if (d.pattern() !== "sequential") throw new Error(`expected sequential, got ${d.pattern()}`);
});

test("pattern random large jumps", () => {
  const d = new AccessPatternDetector();
  for (let i = 0; i < 8; i++) d.observe(i);
  for (let k = 0; k < 12; k++) d.observe(k * 200);
  if (d.pattern() !== "random") throw new Error(`expected random, got ${d.pattern()}`);
});

test("predict_next sequential", () => {
  const d = new AccessPatternDetector();
  for (let i = 0; i < 10; i++) d.observe(i);
  const next = d.predictNext(4, 100);
  if (JSON.stringify(next) !== JSON.stringify([10, 11, 12, 13])) {
    throw new Error(`predict_next: ${JSON.stringify(next)}`);
  }
});

test("coalescePageIndices [3,4,6,7,12] gap=1 → 3 ranges", () => {
  const r = coalescePageIndices([3, 4, 6, 7, 12], 1, DEFAULT_PAGE_SIZE);
  if (r.length !== 3) throw new Error(`expected 3 ranges, got ${r.length}`);
  if (r[0].pageStart !== 3 || r[0].pageEnd !== 4) throw new Error("range 0");
  if (r[1].pageStart !== 6 || r[1].pageEnd !== 7) throw new Error("range 1");
  if (r[2].pageStart !== 12 || r[2].pageEnd !== 12) throw new Error("range 2");
});

test("PageCache LRU evicts at max_pages", () => {
  const c = new PageCache(2, 64);
  c.set(0, new Uint8Array(64));
  c.set(1, new Uint8Array(64));
  c.get(0);
  c.set(2, new Uint8Array(64));
  if (c.has(1)) throw new Error("page 1 should be evicted");
  if (!c.has(0) || !c.has(2)) throw new Error("pages 0 and 2 should remain");
});

await testAsync("InFlightMap dedupes concurrent page loads", async () => {
  const m = new InFlightMap();
  let fetches = 0;
  const p = new Promise((resolve) => {
    setTimeout(() => {
      fetches++;
      resolve(new Uint8Array(8));
    }, 10);
  });
  m.set(3, p);
  const a = m.get(3);
  const b = m.get(3);
  await Promise.all([a, b]);
  if (fetches !== 1) throw new Error(`expected 1 fetch, got ${fetches}`);
});

await testAsync("prefetch_viewport uses ≤3 coalesced fetchRange calls for 50 records", async () => {
  const buf = buildRecords(60);
  const ranges = [];
  const reader = new NxsReader(buf, {
    maxPages: 64,
    coalesceGapPages: 1,
    fetchRange: async (start, len) => {
      ranges.push({ start, len });
      return buf.subarray(start, start + len);
    },
  });
  await reader.prefetch_viewport(0, 49);
  if (ranges.length > 3) {
    throw new Error(`expected ≤3 fetches, got ${ranges.length}: ${JSON.stringify(ranges)}`);
  }
  const stats = reader.cache_stats();
  if (stats.fetches_issued !== ranges.length) {
    throw new Error("fetches_issued mismatch");
  }
});

await testAsync("prefetch_viewport_basic — records readable after prefetch", async () => {
  const buf = buildRecords(55);
  const reader = new NxsReader(buf, { fetchRange: (s, l) => Promise.resolve(buf.subarray(s, s + l)) });
  await reader.prefetch_viewport(0, 49);
  const obj = reader.record(49);
  if (obj.getI64("id") !== 49) throw new Error("record 49 id mismatch");
});

await testAsync("prefetch_memory_eviction", async () => {
  const buf = buildRecords(20);
  const reader = new NxsReader(buf, { maxPages: 2, pageSize: 256, coalesceGapPages: 0 });
  await reader.prefetch_viewport(0, 0);
  await reader.prefetch_viewport(19, 19);
  const stats = reader.cache_stats();
  if (stats.pages_cached > 2) throw new Error(`cache grew past max: ${stats.pages_cached}`);
});

await testAsync("prefetch_deduplication — parallel viewport same page", async () => {
  const buf = buildRecords(10);
  let calls = 0;
  const reader = new NxsReader(buf, {
    maxPages: 8,
    fetchRange: async (s, l) => {
      calls++;
      await new Promise((r) => setTimeout(r, 5));
      return buf.subarray(s, s + l);
    },
  });
  await Promise.all([
    reader.prefetch_viewport(0, 4),
    reader.prefetch_viewport(0, 4),
  ]);
  if (calls > 3) throw new Error(`too many fetches: ${calls}`);
});

await testAsync("hint full small file eager at open", async () => {
  const buf = buildRecords(200);
  const reader = new NxsReader(buf, { hint: HINT_FULL });
  await reader.warmup();
  if (reader.cache_stats().strategy !== "eager") {
    throw new Error(`expected eager strategy, got ${reader.cache_stats().strategy}`);
  }
});

await testAsync("pause stops speculative prefetch", async () => {
  const buf = buildRecords(200);
  const reader = new NxsReader(buf, { autoLifecycle: false });
  for (let i = 0; i < 25; i++) reader.record(i);
  if (reader.cache_stats().pattern !== "sequential") {
    throw new Error(`expected sequential pattern, got ${reader.cache_stats().pattern}`);
  }
  const before = reader.cache_stats().fetches_issued;
  reader.pausePrefetch();
  reader.record(26);
  if (reader.cache_stats().fetches_issued !== before) {
    throw new Error("speculative fetch issued while paused");
  }
  reader.resumePrefetch();
  reader.record(27);
});

await testAsync("sequential upgrade to eager after 150 record() calls", async () => {
  const buf = buildRecords(200);
  const reader = new NxsReader(buf);
  for (let i = 0; i < 150; i++) reader.record(i);
  await reader.warmup();
  const stats = reader.cache_stats();
  if (stats.strategy !== "eager") {
    throw new Error(`expected eager after upgrade, got ${stats.strategy}`);
  }
  if (stats.pattern !== "sequential") {
    throw new Error(`expected sequential pattern, got ${stats.pattern}`);
  }
});

await testAsync("NxsReader.open requires fetch", async () => {
  const orig = globalThis.fetch;
  globalThis.fetch = undefined;
  try {
    let threw = false;
    await NxsReader.open("http://example.com/x.nxb").catch(() => {
      threw = true;
    });
    if (!threw) throw new Error("expected open to fail without fetch");
  } finally {
    globalThis.fetch = orig;
  }
});

console.log(`\n${passed} passed, ${failed} failed\n`);
process.exit(failed > 0 ? 1 : 0);

File: kotlin/src/test/kotlin/nxs/PrefetchTest.kt
package nxs

import kotlin.concurrent.thread
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class PrefetchTest {
    private fun buildRecords(n: Int): ByteArray {
        val schema = NxsSchema(listOf("id", "username", "score", "active"))
        val w = NxsWriter(schema)
        for (i in 0 until n) {
            w.beginObject()
            w.writeI64(0, i.toLong())
            w.writeStr(1, "user_$i")
            w.writeF64(2, i * 0.25)
            w.writeBool(3, i % 2 == 0)
            w.endObject()
        }
        return w.finish()
    }

    private fun buildCompactRecords(n: Int): ByteArray {
        val schema = NxsSchema(listOf("id", "tag"))
        val w = NxsWriter(schema)
        for (i in 0 until n) {
            w.beginObject()
            w.writeI64(0, i.toLong())
            w.writeStr(1, "r$i")
            w.endObject()
        }
        return w.finish()
    }

    @Test
    fun patternUnknownUntilMinObservations() {
        val d = AccessPatternDetector()
        for (i in 0 until 8) d.observe(i)
        assertEquals(AccessPattern.UNKNOWN, d.pattern())
        d.observe(8)
        assertNotEquals(AccessPattern.UNKNOWN, d.pattern())
    }

    @Test
    fun patternSequential_smallDeltas() {
        val d = AccessPatternDetector()
        for (i in 0 until 20) d.observe(i)
        assertEquals(AccessPattern.SEQUENTIAL, d.pattern())
    }

    @Test
    fun patternRandom_largeJumps() {
        val d = AccessPatternDetector()
        for (i in 0 until 8) d.observe(i)
        for (k in 0 until 12) d.observe(k * 200)
        assertEquals(AccessPattern.RANDOM, d.pattern())
    }

    @Test
    fun predictNext_sequential() {
        val d = AccessPatternDetector()
        for (i in 0 until 10) d.observe(i)
        assertTrue(d.predictNext(4, 100).contentEquals(intArrayOf(10, 11, 12, 13)))
    }

    @Test
    fun coalescePageIndices_mergesAdjacentWithGap() {
        val ranges = coalescePageIndices(intArrayOf(3, 4, 6, 7, 12), 1, DEFAULT_PAGE_SIZE)
        assertEquals(3, ranges.size)
        assertEquals(3, ranges[0].pageStart)
        assertEquals(4, ranges[0].pageEnd)
        assertEquals(6, ranges[1].pageStart)
        assertEquals(7, ranges[1].pageEnd)
        assertEquals(12, ranges[2].pageStart)
        assertEquals(12, ranges[2].pageEnd)
        assertEquals(2L * DEFAULT_PAGE_SIZE, ranges[0].byteLength)
    }

    @Test
    fun coalescePageIndices_deduplicates() {
        val ranges = coalescePageIndices(intArrayOf(3, 3, 4), 1, DEFAULT_PAGE_SIZE)
        assertEquals(1, ranges.size)
        assertEquals(3, ranges[0].pageStart)
        assertEquals(4, ranges[0].pageEnd)
    }

    @Test
    fun pageCache_lruEvictsAtMaxPages() {
        val c = PageCache(2, 64)
        c.set(0, ByteArray(64))
        c.set(1, ByteArray(64))
        c.get(0)
        c.set(2, ByteArray(64))
        assertFalse(c.has(1), "page 1 should be evicted")
        assertTrue(c.has(0))
        assertTrue(c.has(2))
    }

    @Test
    fun prefetchViewport_coalescesFetches() {
        val buf = buildRecords(60)
        val ranges = mutableListOf<Pair<Long, Long>>()
        val reader =
            NxsReader(
                buf,
                OpenOptions(
                    maxPages = 64,
                    coalesceGapPages = 1,
                    fetchRange = { off, len ->
                        ranges.add(off to len)
                        buf.copyOfRange(off.toInt(), (off + len).toInt())
                    },
                ),
            )
        reader.prefetchViewport(0, 49)
        assertTrue(ranges.size <= 3, "expected ≤3 fetches, got ${ranges.size}: $ranges")
        assertEquals(ranges.size, reader.cacheStats().fetchesIssued)
    }

    @Test
    fun prefetchViewport_basicRecordsReadable() {
        val buf = buildRecords(55)
        val reader =
            NxsReader(
                buf,
                OpenOptions(fetchRange = { off, len -> buf.copyOfRange(off.toInt(), (off + len).toInt()) }),
            )
        reader.prefetchViewport(0, 49)
        assertEquals(49L, reader.record(49).getI64("id"))
    }

    @Test
    fun prefetchMemoryEviction() {
        val buf = buildRecords(20)
        val reader =
            NxsReader(
                buf,
                OpenOptions(maxPages = 2, pageSize = 256, coalesceGapPages = 0),
            )
        reader.prefetchViewport(0, 0)
        reader.prefetchViewport(19, 19)
        assertTrue(reader.cacheStats().pagesCached <= 2)
    }

    @Test
    fun prefetchDeduplication_concurrentViewport() {
        val buf = buildRecords(10)
        var calls = 0
        val reader =
            NxsReader(
                buf,
                OpenOptions(
                    maxPages = 8,
                    fetchRange = { off, len ->
                        synchronized(this) { calls++ }
                        Thread.sleep(5)
                        buf.copyOfRange(off.toInt(), (off + len).toInt())
                    },
                ),
            )
        val t1 = thread { reader.prefetchViewport(0, 4) }
        val t2 = thread { reader.prefetchViewport(0, 4) }
        t1.join()
        t2.join()
        assertTrue(calls <= 3, "too many fetches: $calls")
    }

    @Test
    fun openOptions_hintStored() {
        val buf = buildRecords(5)
        val reader = NxsReader(buf, OpenOptions(hint = AccessHint.SEQUENTIAL))
        assertEquals(AccessHint.SEQUENTIAL, reader.accessHint())
    }

    @Test
    fun cacheStats_defaults() {
        val buf = buildRecords(5)
        val reader = NxsReader(buf)
        val stats = reader.cacheStats()
        assertEquals(DEFAULT_MAX_PAGES, stats.pagesMax)
        assertEquals("adaptive", stats.strategy)
        assertEquals("unknown", stats.pattern)
    }

    @Test
    fun sequentialUpgrade_toEagerAfter150Records() {
        val buf = buildCompactRecords(200)
        val reader = NxsReader(buf)
        try {
            for (i in 0 until 150) reader.record(i)
            reader.warmup()
            assertEquals("eager", reader.cacheStats().strategy)
            assertEquals("sequential", reader.cacheStats().pattern)
        } finally {
            reader.close()
        }
    }

    @Test
    fun pauseStopsSpeculative() {
        val buf = buildCompactRecords(200)
        val reader = NxsReader(buf)
        try {
            for (i in 0 until 25) reader.record(i)
            assertEquals("sequential", reader.cacheStats().pattern)
            val before = reader.cacheStats().fetchesIssued
            reader.pausePrefetch()
            reader.record(26)
            assertEquals(before, reader.cacheStats().fetchesIssued)
            reader.resumePrefetch()
            reader.record(27)
            assertTrue(reader.cacheStats().fetchesIssued >= before)
        } finally {
            reader.close()
        }
    }

    @Test
    fun hintFull_eagerAtOpen() {
        val buf = buildCompactRecords(200)
        val reader = NxsReader(buf, OpenOptions(hint = AccessHint.FULL))
        try {
            reader.warmup()
            assertEquals("eager", reader.cacheStats().strategy)
        } finally {
            reader.close()
        }
    }

    @Test
    fun prefetchColumn_singleFetchBeforeColSum() {
        val candidates =
            listOf(
                java.nio.file.Paths.get("../../nyxis/conformance/columnar_flat8_dense_100.nxb"),
                java.nio.file.Paths.get("../conformance/columnar_flat8_dense_100.nxb"),
            )
        val path = candidates.firstOrNull { java.nio.file.Files.isRegularFile(it) } ?: return
        val data = java.nio.file.Files.readAllBytes(path)
        var fetches = 0
        val reader =
            NxsReader(
                data,
                OpenOptions(fetchRange = { off, len ->
                    fetches++
                    data.copyOfRange(off.toInt(), (off + len).toInt())
                }),
            )
        try {
            reader.prefetchColumn("score")
            assertEquals(1, fetches)
            val sum = reader.sumF64("score")
            reader.prefetchColumn("score")
            assertEquals(1, fetches)
            assertEquals(2475.0, sum, 1e-6)
            assertEquals(1, reader.cacheStats().columnFetchesIssued)
        } finally {
            reader.close()
        }
    }
}
